src/core/public/context/context.ts

```diff
@@ -0,0 +1,281 @@
+ *       },
+ *       registerRoute(plugin, path, handler) {
+ *         this.handlers.set(
+ *           path,
+ *           // the handler will not receive any contexts provided by other dependencies of the calling plugin.
+ *           this.contextContainer.createHandler('my_plugin', handler)
+ *         );
+ *       }
+ *     }
+ *   }
+ * }
+ * ```
+ *
+ * @public
+ */
+export interface ContextContainer<
+  TContext extends {},
+  THandlerReturn,
+  THandlerParameters extends any[] = []
+> {
+  /**
+   * Register a new context provider. Throws an exception if more than one provider is registered for the same
+   * `contextName`.
+   *
+   * @param plugin - The plugin ID for the plugin that registers this context.
+   * @param contextName - The key of the `TContext` object this provider supplies the value for.
+   * @param provider - A {@link ContextProvider} to be called each time a new context is created.
+   * @returns The `ContextContainer` for method chaining.
+   */
+  registerContext<TContextName extends keyof TContext>(
+    plugin: string | CoreId,
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, THandlerParameters>
+  ): this;
+
+  /**
+   * Create a new handler function pre-wired to context for the plugin.
+   *
+   * @param plugin - The plugin ID for the plugin that registers this context.
+   * @param handler
+   * @returns A function that takes `THandlerParameters`, calls `handler` with a new context, and returns a Promise of
+   * the `handler` return value.
+   */
+  createHandler(
+    plugin: string | CoreId,
+    handler: Handler<TContext, THandlerReturn, THandlerParameters>
+  ): (...rest: THandlerParameters) => Promisify<THandlerReturn>;
+}
+
+type ContextSource = PluginName | CoreId;
+
+/** @internal */
+export class ContextContainerImplementation<
+  TContext extends {},
+  THandlerReturn,
+  THandlerParameters extends any[] = []
+> implements ContextContainer<TContext, THandlerReturn, THandlerParameters> {
+  /**
+   * Used to map contexts to their providers and associated plugin. In registration order which is tightly coupled to
+   * plugin load order.
+   */
+  private readonly contextProviders = new Map<
+    keyof TContext,
+    {
+      provider: ContextProvider<TContext, keyof TContext, THandlerParameters>;
+      source: ContextSource;
+    }
+  >();
+  /** Used to keep track of which plugins registered which contexts for dependency resolution. */
+  private readonly contextNamesBySource: Map<ContextSource, Array<keyof TContext>>;
+
+  /**
+   * @param pluginDependencies - A map of plugins to an array of their dependencies.
+   */
+  constructor(
+    private readonly pluginDependencies: ReadonlyMap<PluginName, PluginName[]>,
+    private readonly coreId: CoreId
+  ) {
+    this.contextNamesBySource = new Map<ContextSource, Array<keyof TContext>>([[coreId, []]]);
+  }
+
+  public registerContext = <TContextName extends keyof TContext>(
```

Comment:
Can you elaborate on why we are using arrow functions here instead of class methods?

Reply:
Without using function properties, I was having issues with `this` being the caller. I'll play around with this more to see if it's necessary.